pkg/endpoint/endpoint.go

```diff
@@ -1458,8 +1458,26 @@ func (e *Endpoint) OnProxyPolicyUpdate(revision uint64) {
 	e.unlock()
 }
 
-// OnDNSPolicyUpdateLocked is called when the Endpoint's DNS policy has been updated
-func (e *Endpoint) OnDNSPolicyUpdateLocked(rules restore.DNSRules) {
+// UpdateDNSRulesIfNeeded is called to synchronize the Endpoint's DNS rules
+// with what stored inside DNS proxy for this endpoint.
+func (e *Endpoint) UpdateDNSRulesIfNeeded() {
+	e.unconditionalLock()
```

Comment:
The previous callers did a `lockAlive()`, which says that it is attempting to handle the case where the caller is calling this function after the endpoint was deleted. However, this `e.unconditionalLock()` assumes that the `Endpoint` *must* be still valid when the function is called. The callers are still using `e.rLockAlive()` calls after the call to this new function though, which again attempts to handle the case where the endpoint may have been deleted. 

I don't exactly know the callpaths to tell which of these is safe, though I'm immediately suspicious since it's called from a trigger function[^1], hence could be delayed by ~seconds from the actual `Trigger()` call. During this period, the endpoint could get deleted. I suspect that the simple answer is just to change this function to use `e.lockAlive()` and return an error, so that this call can also potentially exit early if the endpoint is no longer valid.

[^1]: `Endpoint.SyncEndpointHeaderfile()` -> `Endpoint.syncEndpointHeaderfile()` -> `Endpoint.UpdateDNSRulesIfNeeded()`